bots/testinfo/src/main/java/org/openjdk/skara/bots/testinfo/TestInfoBot.java

```diff
@@ -31,40 +31,37 @@
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.logging.Logger;
 
+/**
+ * The TestInfoBot copies 'checks' from the source repository of a PR to the
+ * PR itself. In GitHub, these checks are usually workflow/action runs which
+ * users may have activated on their personal forks. By copying them to a PR,
+ * reviewers can easily see the status of the last workflow runs directly in
+ * the PR.
+ * <p>
+ * The bot polls for work using the standard PullRequestPoller, so will
+ * process any updated PR. Depending on the outcome of this processing, the
+ * TestInfoBotWorkItem calls back to the bot with a re-check request which
+ * causes the bot to submit that PR again after the specified amount of time,
+ * or earlier if another change the PR has been detected.
+ */
 public class TestInfoBot implements Bot {
     private final HostedRepository repo;
-    private final Map<String, Instant> expirations = new ConcurrentHashMap<>();
-
-    private static final Logger log = Logger.getLogger("org.openjdk.skara.bots");
+    private final PullRequestPoller poller;
 
     TestInfoBot(HostedRepository repo) {
         this.repo = repo;
-    }
-
-    private String pullRequestToKey(PullRequest pr) {
-        return pr.id() + "#" + pr.headHash().hex();
+        this.poller = new PullRequestPoller(repo, true);
     }
 
     @Override
     public List<WorkItem> getPeriodicItems() {
-        var prs = repo.pullRequestsAfter(ZonedDateTime.now().minus(Duration.ofDays(1)));
-        var ret = new ArrayList<WorkItem>();
-        for (var pr : prs) {
-            if (pr.sourceRepository().isEmpty()) {
-                continue;
-            }
-
-            var expirationKey = pullRequestToKey(pr);
-            if (expirations.containsKey(expirationKey)) {
-                var expiresAt = expirations.get(expirationKey);
-                if (expiresAt.isAfter(Instant.now())) {
-                    continue;
-                }
-            }
-
-            ret.add(new TestInfoBotWorkItem(pr, expiresIn -> expirations.put(expirationKey, Instant.now().plus(expiresIn))));
-        }
-        return ret;
+        var now = Instant.now();
+        var prs = poller.updatedPullRequests();
+        return prs.stream()
+                .filter(pr -> pr.sourceRepository().isPresent())
+                .map(pr -> (WorkItem) new TestInfoBotWorkItem(pr,
+                        delay -> poller.retryPullRequest(pr, now.plus(delay))))
```

Comment:
Maybe I'm getting lost here, but this feels wrong to me. The `now` is computed as the now of the call to `getPeriodicItems()`, but it is used in the lambda for the `retry` operation of `TestInfoBotWorkItem`. But that retry should happen `delay` time after it is evaluated, right? 

It might not have much practical impact, but it just looks odd to me, like mixing different time streams.

Reply:
To expand a bit: the `now.plus(delay)` will be evaluated in the `run` method of the work item, but then the result will be offset not from the time of `run` but from the time of the call to this method. What would happen if it, for some reason, would take more than 2 minutes between these two events, and we'd get a time in the past. Will it be retried immediately, or will it cause trouble with the poller? 

And also it feel conceptually wrong; the retry time feels like it should be specified from when the decision is made for the next check.

Reply:
You are right, that is odd, it really should be X minutes after it was evaluated, not after the `getPeriodicItems` call. When bots are busy, the time difference can easily be much longer than 2 minutes, and in that case, effectively ignoring the delay is not helping in reducing the load.

The poller should handle past time fine though, it will basically be equivalent to a retry without a time component.